Taxonomy reorg per dewey decimal classifications

mcorbin-ibm opened the pull request:
Reorganized the taxonomy domains and subdomains to align with the Dewey Decimal Classifications

bjhargrave reviewed (commented):
Nice!

If we could, I would prefer to keep compositional_skills and knowledge folders free of readme files. That is, any change in these trees _are_ a contribution to the taxonomy rather than some doc improvements. The current readme in knowledge is annoying in this way :-)

mcorbin-ibm commented:
@bjhargrave 

> If we could, I would prefer to keep compositional_skills and knowledge folders free of readme files. That is, any change in these trees _are_ a contribution to the taxonomy rather than some doc improvements. The current readme in knowledge is annoying in this way :-)

Do you mean just the main parent folders (knowledge, compositional_skills, and foundational_skills)?  OR, do you not want readme files in the domain/subdomain folders?

We do have a docs folder, where we could put most of the info for the repo, to keep the taxonomy tree folders clear?

I do still need to "document" the taxonomy tree -- which I was going to do in readme.txt files for the domains/subdomains??

bjhargrave commented:
> Do you mean just the main parent folders (knowledge, compositional_skills, and foundational_skills)? OR, do you not want readme files in the domain/subdomain folders?

My request would to not have readme files anywhere under the taxonomy folders (knowledge, compositional_skills, and foundational_skills). 

> I do still need to "document" the taxonomy tree -- which I was going to do in readme.txt files for the domains/subdomains??

Agree but I would rather see that all together in a single readme file since that would give the reader a broad view over the taxonomy organization rather than walking around the tree encountering readme files occasionally.

mcorbin-ibm commented:
@jjasghar I have removed the readme files, and updated the main repo's readme file for these changes.  There might be some additional changes to verify the qna.yaml files are either "grounded" or "ungrounded" and putting them in those subfolders.  And, when we do start merging other knowledge contributions, we need to remember to add the document_type as the final node in the tree.  I couldn't quickly find/identify any qna.yaml files to do that with.  Please review my latest change here, and I'll let you do the honors of removing DRAFT!  :)

jjasghar commented on compositional_skills/linguistics/writing/prose/debate/ungrounded/qna.yaml:
We should discuss this in the triage meeting. I'm not _sure_ how we can create 5 items on this one.

bjhargrave replied:
There are a number of skills which ideally need to be updated to 5 examples...

jjasghar commented:
@bjhargrave can you confirm https://github.com/instructlab/taxonomy/actions/runs/9751722026/job/26913897281?pr=1215 that is the same thing as the /files directory?

bjhargrave commented on foundational_skills/linguistics/logical_sequence_of_words/qna.yaml:
I am not sure we should touch foundational_skills folder at all? Need to get confirmation from backend team.

mcorbin-ibm replied:
I talked to Kai, and tried to get him to explain to me how "linguistics_reasoning" was different from "linguistics" as a foundational skill in itself.

He said:  "Linguistics reasoning is: 'it's doing a reasoning task in the form of natural language, like "if Alice is Ben's mon and Cole is Ben's dad, what does Alice call Cole"' AND linguistics ability is:  "the basics of using language to make a conversation, understand context/instructions, etc. etc."

I'm not sure I get the distinction, but if there is a technical reason we should just leave foundational_skills alone, then we can keep the linguistics_reasoning folder and its subfolders where they were originally.

bjhargrave commented on README.md:
```diff
@@ -28,6 +28,14 @@ method.
 
+## Choosing domains for the taxonomy
+
+In general, we use the Dewey Decimal Classification (DDC) System to determine our domains (and subdomains) in the taxonomy. This SUMMARIES of the DDC document is a great resourcce for determining where a topic might be classified:
```
```suggestion
In general, we use the Dewey Decimal Classification (DDC) System to determine our domains (and subdomains) in the taxonomy. The [DDC SUMMARIES document](https://www.oclc.org/content/dam/oclc/dewey/resources/summaries/deweysummaries.pdf) is a great resource for determining where a topic might be classified.
```

bjhargrave commented on README.md:
```diff
@@ -28,6 +28,14 @@ method.
+
+In general, we use the Dewey Decimal Classification (DDC) System to determine our domains (and subdomains) in the taxonomy. This SUMMARIES of the DDC document is a great resourcce for determining where a topic might be classified:
+
+https://www.oclc.org/content/dam/oclc/dewey/resources/summaries/deweysummaries.pdf
```
```suggestion
```

bjhargrave commented on README.md:
```diff
@@ -28,6 +28,14 @@ method.
+
+https://www.oclc.org/content/dam/oclc/dewey/resources/summaries/deweysummaries.pdf
+
+If you are unsure where to put your knowledge or compositional skill, create a folder in the `miscellaneous_unknown` folder.
```
```suggestion
If you are unsure where to put your knowledge or compositional skill, create a folder in the `miscellaneous_unknown` folder somewhere under the `knowledge` or `compositional_skills` folders.
```

bjhargrave commented on README.md:
```diff
@@ -53,6 +61,8 @@ Your skills contribution pull requests must include the following:
 > There is a limit to how much content can exist in the question/answer pairs for the model to process. Due to this, only add a maximum
 > of around 2300 words to your question and answer seed example pairs in the `qna.yaml` file.
 
+Compositional skills can either be grounded (includes a context) or ungrounded (does not include a context).  Grounded or ungrounded is the last node in the taxonomy tree, for example: `linguistics/writing/poetry/haiku/ungrounded` or `linguistics/grammar/grounded`. The `qna.yaml` is in this final node.
```
Again, I am unclear on the utility of putting grounded/ungrounded in the path (since the qna.yaml can have or not have context independent of the path).

bjhargrave commented on README.md:
```diff
@@ -245,8 +257,9 @@ seed_examples:
-data, or references.
+While skills are foundational or performative, knowledge is based more on answering questions that involve facts, data, or references.
+
+Knowledge is supported by documents, such as textbook, technical_manual, encyclopedia, journal, or magazine. The document type is the last node in the taxonomy tree, for example: `science/animals/tabby_cat/encyclopedia`. The `qna.yaml` is in this final node.
```
```suggestion
Knowledge is supported by documents, such as a textbook, technical manual, encyclopedia, journal, or magazine. The document type is the last node in the taxonomy tree, for example: `science/animals/tabby_cat/encyclopedia`. The `qna.yaml` is in this final node.
```

Also, not sure of the utility of putting the doc type in the path. It is also possible the knowledge documents come from multiple sources: web page, pdf document, wikipedia article.

mcorbin-ibm replied:
@bjhargrave This was previously discussed and agreed upon that it would be the final node in the path.  I would very much like to see this JUST in the qna.yaml file in the document: section, but the majority of folks said that they'd like to see it in the taxonomy tree for when they just scan/read the tree.

bjhargrave replied:
I would like to open this up for discussion again :-)

mcorbin-ibm commented on compositional_skills/linguistics/grammar/basic_grammer_tests/ungrounded/qna.yaml:
@bjhargrave Because we added "doc_type" at the trailing end of the path, we added "grounded" or "ungrounded" at the trailing end of the path.  Same concept applies:  If we want to expose the doc_type in the tree, we should expose the notion of a context being provided.  Everyone seemed to think that doc_type should be included in the path, so that is why I kept grounded/ungrounded in a similar location (previously, it appeared at the beginning, separating domains into grounded or freeform, which didn't make sense to me.

bjhargrave replied:
It is a place where things can be out-of-sync. That is, I can use a path ending in `ungrounded` and have `context` elements in my `qna.yaml`. So this can be misleading. We can also end up with `a/b/ungrounded/qna.yaml` and `a/b/grounded/qna.yaml` for the same `a/b` skill (this was true before but if we are changing things now, we can try to avoid it).

makelinux commented:
I recognize that unambiguous classification is an extremely complex task. Here are some of my thoughts:

- Often, a topic belongs to multiple categories. For example, an **electric battery** can be classified under **chemistry** from a production standpoint, **physics** based on its function, and **electronics** based on its usage.
- In classification, it is essential to determine which categories are top-level and which are subcategories.
- Looking at the top categories of the Dewey Decimal Classification (DDC), in my opinion, 'science' is merely a form of knowledge and should not be a top-level category.
- Consider trying to classify **cooking** and **culinary** in DDC. I couldn’t find a suitable category.  Can you? _Arts and recreation?_
- Generally, it seems to me that DDC is more suited to a scientific and academic perspective on printed books and less applicable to all forms of knowledge.

mcorbin-ibm commented:
@makelinux 

> I recognize that unambiguous classification is an extremely complex task. 
Indeed!  :)

> * Often, a topic belongs to multiple categories. For example, an **electric battery** can be classified under **chemistry** from a production standpoint, **physics** based on its function, and **electronics** based on its usage.

Yes, you can always classify topics into different categories, and I think much will depend upon the specific knowledge being submitted.  If the knowledge is talking about its function, then classifying it under physics might be best, but if the knowledge is talking about where batteries are used, then maybe it belongs in technology/electronics.  I'm not sure that there is a way around this and it is just something that we will have to make a judgement call as to where a piece of knowledge belongs.

> * In classification, it is essential to determine which categories are top-level and which are subcategories.
> * Looking at the top categories of the Dewey Decimal Classification (DDC), in my opinion, 'science' is merely a form of knowledge and should not be a top-level category.
> * Generally, it seems to me that DDC is more suited to a scientific and academic perspective on printed books and less applicable to all forms of knowledge.

The DDC has been around for nearly 150 years, and is in its 20th edition.  It has 10 top categories, subdivided in the 100s, subdivided again to the 1000s.  Please see this summaries doc:  https://www.oclc.org/content/dam/oclc/dewey/resources/summaries/deweysummaries.pdf.  It is meant to be a standard classification of all forms of knowledge.  And, we will certainly run into some cases where we will have to find a "best fit" for a knowledge, but starting from the top 10 categories and its 10 subcategories seemed to present the best starting point.

As a secondary source to help us with classification of knowledge, we can look to Wikipedia to see how/where they placed things or to get additional ideas:  https://en.wikipedia.org/wiki/Wikipedia:Contents.

The InstructLab taxonomy will not be a direct 1:1 mapping of the DDC, but the starting point to finding a best fit for the topics of knowledge.

> * Consider trying to classify **cooking** and **culinary** in DDC. I couldn’t find a suitable category.  Can you? _Arts and recreation?_

For cooking and "culinary arts" I would put them in technology/food_and_drink.

bjhargrave commented on README.md:
```diff
@@ -53,6 +59,8 @@ Your skills contribution pull requests must include the following:
 > There is a limit to how much content can exist in the question/answer pairs for the model to process. Due to this, only add a maximum
 > of around 2300 words to your question and answer seed example pairs in the `qna.yaml` file.
 
+Compositional skills can either be grounded (includes a context) or ungrounded (does not include a context).  Grounded or ungrounded is the last node in the taxonomy tree, for example: `linguistics/writing/poetry/haiku/ungrounded` or `linguistics/grammar/grounded`. The `qna.yaml` is in this final node.
```
We need to strike the reference to ungrounded in the path here.

jjasghar replied:
updated.

bjhargrave commented on README.md:
```diff
@@ -61,7 +69,7 @@ Taxonomy skill files must be a valid [YAML](https://yaml.org/) file named `qna.y
   submissions should have at least five entries, although
   older files may have fewer. **Required**
-  - `context`: Grounded skills require the user to provide context containing information that the model is expected to take into account during processing. This is different from knowledge, where the model is expected to gain facts and background knowledge from the tuning process. The context key is optional for freeform skills.
+  - `context`: Grounded skills require the user to provide context containing information that the model is expected to take into account during processing. This is different from knowledge, where the model is expected to gain facts and background knowledge from the tuning process. The context key is optional for ungrounded skills.
```
```suggestion
  - `context`: Grounded skills require the user to provide context containing information that the model is expected to take into account during processing. This is different from knowledge, where the model is expected to gain facts and background knowledge from the tuning process. The context key should not be used for ungrounded skills.
```

bjhargrave commented on README.md:
```diff
@@ -149,23 +157,25 @@ Here is the location of this YAML in the taxonomy tree. Note that the YAML file
-    |   |   |        attribution.txt
+    └── poetry
+    |   └── haiku <=== here it is :)
+    |   |   └── ungrounded
```
No ungrounded

bjhargrave commented on README.md:
```diff
@@ -149,23 +157,25 @@ Here is the location of this YAML in the taxonomy tree. Note that the YAML file
+        [...]
+    └── prose
+    |   └── debate
+    |   |   └── ungrounded
```
No ungrounded

bjhargrave commented on README.md:
```diff
@@ -245,8 +255,9 @@ seed_examples:
-data, or references.
+While skills are foundational or performative, knowledge is based more on answering questions that involve facts, data, or references.
+
+Knowledge is supported by documents, such as a textbook, technical manual, encyclopedia, journal, or magazine. The document type is the last node in the taxonomy tree, for example: `science/animals/tabby_cat/encyclopedia`. The `qna.yaml` is in this final node.
```
```suggestion
Knowledge is supported by documents, such as a textbook, technical manual, encyclopedia, journal, or magazine.
```

bjhargrave commented on README.md:
```diff
@@ -571,48 +582,38 @@ each branch, there is a YAML file (qna.yaml) that contains the examples for that
     │   │   ├── idea_generation
-    │   │   │   └── qna.yaml
-    │   │   │       attribution.txt
+    |   |   |   └── ungrounded
```
no ungrounded

bjhargrave commented on README.md:
```diff
@@ -571,48 +582,38 @@ each branch, there is a YAML file (qna.yaml) that contains the examples for that
     │   │   ├── refute_claim
-    │   │   │   └── qna.yaml
-    │   │   │       attribution.txt
+    |   |   |   └── ungrounded
```
no ungrounded

bjhargrave commented on README.md:
```diff
@@ -571,48 +582,38 @@ each branch, there is a YAML file (qna.yaml) that contains the examples for that
-    │   │   │   │   └── qna.yaml
-    │   │   │   │       attribution.txt
-    │   │   │   └── informal
+    |   |   |   └── ungrounded
```
no ungrounded

bjhargrave commented:
I have force pushed changes which include making .gitignore files empty. 

Please don't push any merge commits to the PR.

jjasghar commented:
I think it's ready to merge!

https://www.youtube.com/watch?v=NHiUQb5xg7A
